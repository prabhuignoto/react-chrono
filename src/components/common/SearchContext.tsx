import React, { createContext, useState, useContext } from 'react';

interface SearchContextType {
  searchTerm: string;
  setSearchTerm: (term: string) => void;
}

// Create a context with default values
const SearchContext = createContext<SearchContextType>({
  searchTerm: '',
  setSearchTerm: () => {},
});

/**
 * Provider component for the search context
 */
export const SearchProvider: React.FC<{ children: React.ReactNode }> = ({
  children,
}) => {
  const [searchTerm, setSearchTerm] = useState<string>('');

  return (
    <SearchContext.Provider value={{ searchTerm, setSearchTerm }}>
      {children}
    </SearchContext.Provider>
  );
};

/**
 * Custom hook to use the search context
 */
export const useSearch = (): SearchContextType => {
  const context = useContext(SearchContext);

  if (context === undefined) {
    throw new Error('useSearch must be used within a SearchProvider');
  }

  return context;
};

export default SearchContext;
