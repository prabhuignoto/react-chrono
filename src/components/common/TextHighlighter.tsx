import React from 'react';
import styled from 'styled-components';
import { Theme } from '@models/Theme';

interface HighlightedTextProps {
  text: string;
  searchTerm: string;
  theme: Theme;
}

// Styled component for highlighted text
const HighlightSpan = styled.span<{ theme: Theme }>`
  background-color: ${({ theme }) =>
    theme.primary}33; /* Primary color with 20% opacity */
  color: ${({ theme }) => theme.primary};
  font-weight: 500;
  border-radius: 2px;
  padding: 0 2px;
`;

/**
 * Component that highlights occurrences of a search term within text
 */
const TextHighlighter: React.FC<HighlightedTextProps> = ({
  text,
  searchTerm,
  theme,
}) => {
  // If no search term or no text, just return the original text
  if (!searchTerm || !text) {
    return <>{text}</>;
  }

  // Escape special regex characters in the search term
  const escapedSearchTerm = searchTerm.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');

  // Create a case-insensitive regular expression to find all occurrences
  const regex = new RegExp(`(${escapedSearchTerm})`, 'gi');

  // Split the text by the search term
  const parts = text.split(regex);

  // Return the text with highlighted parts
  return (
    <>
      {parts.map((part, i) =>
        regex.test(part) ? (
          <HighlightSpan key={i} theme={theme}>
            {part}
          </HighlightSpan>
        ) : (
          <React.Fragment key={i}>{part}</React.Fragment>
        ),
      )}
    </>
  );
};

export default TextHighlighter;
