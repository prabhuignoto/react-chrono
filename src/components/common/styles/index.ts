import { Theme } from '@models/Theme';
import { css } from 'styled-components';

export const ScrollBar = css<{ theme: Theme }>`
  scrollbar-color: ${(p) => p.theme?.primary} default;
  scrollbar-width: thin;

  &::-webkit-scrollbar {
    width: 0.3em;
  }

  &::-webkit-scrollbar-track {
    box-shadow: inset 0 0 6px rgba(0, 0, 0, 0.2);
  }

  &::-webkit-scrollbar-thumb {
    background-color: ${(p) => p.theme?.primary};
    outline: 1px solid ${(p) => p.theme?.primary};
  }
`;
