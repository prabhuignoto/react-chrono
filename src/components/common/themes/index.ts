import { Theme } from '@models/Theme';

export const defaultTheme: Theme = {
  cardBgColor: '#ffffff',
  cardDetailsBackGround: '#ffffff',
  cardDetailsColor: '#000',
  cardMediaBgColor: '#f5f5f5',
  cardSubtitleColor: '#000',
  cardTitleColor: '#007FFF',
  detailsColor: '#000',
  iconBackgroundColor: '#007FFF',
  nestedCardBgColor: '#f5f5f5',
  nestedCardDetailsBackGround: '#f5f5f5',
  nestedCardDetailsColor: '#000',
  nestedCardSubtitleColor: '#000',
  nestedCardTitleColor: '#000',
  primary: '#007FFF',
  secondary: '#ffdf00',
  titleColor: '#007FFF',
  titleColorActive: '#007FFF',
  toolbarBgColor: '#f5f5f5',
  toolbarBtnBgColor: '#fff',
  toolbarTextColor: '#000',
};

export const darkTheme: Theme = {
  cardBgColor: '#191919',
  cardDetailsBackGround: '#191919',
  cardDetailsColor: '#ffff0f',
  cardMediaBgColor: '#2f2f2f',
  cardSubtitleColor: '#ffffff',
  cardTitleColor: '#007FFF',
  detailsColor: '#ffffff',
  iconBackgroundColor: '#007FFF',
  nestedCardBgColor: '#333333',
  nestedCardDetailsBackGround: '#333333',
  nestedCardDetailsColor: '#ffffff',
  nestedCardSubtitleColor: '#ffffff',
  nestedCardTitleColor: '#ffffff',
  primary: '#007FFF',
  secondary: '#ffdf00',
  titleColor: '#007FFF',
  titleColorActive: '#007FFF',
  toolbarBgColor: '#555',
  toolbarBtnBgColor: '#222',
  toolbarTextColor: '#ffffff',
};
