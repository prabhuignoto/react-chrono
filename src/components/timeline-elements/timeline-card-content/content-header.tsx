import { FunctionComponent, memo, useContext, useMemo } from 'react';
import { GlobalContext } from '../../GlobalContext';
import { SubTitleMemo } from '../memoized/subtitle-memo';
import { TitleMemo } from '../memoized/title-memo';
import { ContentHeaderProps } from './header-footer.model';
import { CardTitle, TimelineCardHeader } from './timeline-card-content.styles';

/**
 * ContentHeader component
 * This component renders the header of the timeline card including the title and subtitle.
 * It doesn't render the title and subtitle if the card has media.
 * The title and subtitle are memoized to prevent unnecessary re-renders.
 *
 * @property {string} title - The title of the card.
 * @property {string} url - The URL of the card.
 * @property {boolean} media - Indicates whether the card has media or not.
 * @property {string} content - The main content of the card.
 * @returns {JSX.Element} The ContentHeader component.
 */
const ContentHeader: FunctionComponent<ContentHeaderProps> = memo(
  ({ title, url, media, content, cardTitle }: ContentHeaderProps) => {
    // Using context to get global values
    const { fontSizes, classNames, theme, isMobile } =
      useContext(GlobalContext);

    const isNotMedia = useMemo(() => !media, [media]);

    return (
      <TimelineCardHeader>
        {/* Render title if there is no media */}

        {isMobile ? (
          <CardTitle $fontSize={'1.2rem'} theme={theme}>
            {cardTitle}
          </CardTitle>
        ) : null}

        {isNotMedia ? (
          <TitleMemo
            title={title}
            theme={theme}
            url={url}
            fontSize={fontSizes?.cardTitle}
            classString={classNames?.cardTitle}
          />
        ) : null}

        {/* Render subtitle if there is no media */}

        {isNotMedia ? (
          <SubTitleMemo
            content={content}
            theme={theme}
            fontSize={fontSizes?.cardSubtitle}
            classString={classNames?.cardSubTitle}
          />
        ) : null}
      </TimelineCardHeader>
    );
  },
);

// Setting display name for easier debugging
ContentHeader.displayName = 'ContentHeader';

export { ContentHeader };
