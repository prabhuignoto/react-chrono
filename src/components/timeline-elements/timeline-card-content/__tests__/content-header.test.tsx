import { describe, expect, it } from 'vitest';
import { customRender } from '../../../common/test';
import { providerProps } from '../../../common/test/index';
import { ContentHeader } from '../content-header';

describe('Content Header', () => {
  // should render the title and subtitle of the card
  it('should render the title and subtitle of the card', () => {
    const { getByText, container } = customRender(
      <ContentHeader
        title="title"
        content="content"
        data-testid="header-test-1"
      />,
      {
        providerProps: {
          ...providerProps,
          media: null,
        },
      },
    );

    const titleElement = container.querySelector('.card-title');
    expect(titleElement).toBeInTheDocument();
    expect(titleElement).toHaveTextContent('title');
    expect(getByText('content')).toBeInTheDocument();
  });

  // should match the snapshot
  it('should match the snapshot', () => {
    const { container } = customRender(
      <ContentHeader
        title="title"
        content="content"
        data-testid="header-test-2"
      />,
      {
        providerProps,
      },
    );

    expect(container).toMatchSnapshot();
  });

  // should render the link if url is provided
  it('should render the link if url is provided', () => {
    const { getByRole, container } = customRender(
      <ContentHeader
        title="title"
        content="content"
        url="http://www.google.com"
        data-testid="header-test-3"
      />,
      {
        providerProps,
      },
    );

    const titleElement = container.querySelector('.card-title');
    expect(titleElement).toBeInTheDocument();
    expect(titleElement).toHaveTextContent('title');
    expect(getByRole('link')).toHaveAttribute('href', 'http://www.google.com');
  });
});
