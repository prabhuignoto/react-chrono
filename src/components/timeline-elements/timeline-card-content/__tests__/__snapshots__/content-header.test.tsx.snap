// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Content Header > should match the snapshot 1`] = `
<div>
  <header
    class="sc-jsEeTM jtVkFt"
  >
    <h1
      class="sc-irLvIq cwgmmW card-title"
      data-class="card-title"
    >
      title
    </h1>
    <h2
      class="sc-kFCroH lfQnzg card-sub-title card-subtitle"
    >
      content
    </h2>
  </header>
</div>
`;
